### PR TITLE
Grad accum

### PR DESCRIPTION
**Pull request by dimakarp1996**
Added gradient accumulation. Tested on SST-2.

### CODE REVIEW

File: deeppavlov/models/bert/bert_classifier.py [hunk 1]
```diff
@@ -200,27 +201,91 @@ def _build_feed_dict(self, input_ids, input_masks, token_types, y=None):
...
+        assert num_parts <= len(features)
+        num_features = math.ceil(len(features) / num_parts)
+        feature_batches = [features[i:i + num_features] for i in range(0, len(features), num_features)]
+        return feature_batches
+
+    def train_on_batch_old(self, features: List[InputFeatures], y: Union[List[int], List[List[int]]]) -> Dict:
         """Train model on given batch.
         This method calls train_op using features and y (labels).
-
         Args:
```
Comment (dilyararimovna) [inline deeppavlov/models/bert/bert_classifier.py:221]: эти пустые строки в док стрингах не только для красоты, но и для корректной конвертации в readthedocs. верни все пустые строки на место

File: deeppavlov/core/trainers/utils.py [hunk 2]
```diff
@@ -17,7 +17,6 @@
 from typing import List, Tuple, Union, Iterable
 
 import numpy
-
 from deeppavlov.core.common.metrics_registry import get_metric_by_name
```
Comment (dilyararimovna) [inline deeppavlov/core/trainers/utils.py:20]: верни пустую строку. здесь правила для импортов

Comment (dilyararimovna) [reply deeppavlov/core/trainers/utils.py:20]: и дальше в файлах сделай правильный порядок импортов с пустыми строками

File: deeppavlov/models/bert/bert_classifier.py [hunk 3]
```diff
@@ -200,27 +201,91 @@ def _build_feed_dict(self, input_ids, input_masks, token_types, y=None):
...
         input_masks = [f.input_mask for f in features]
         input_type_ids = [f.input_type_ids for f in features]
 
         feed_dict = self._build_feed_dict(input_ids, input_masks, input_type_ids, y)
 
-        _, loss = self.sess.run([self.train_op, self.loss], feed_dict=feed_dict)
+        loss = self.sess.run(self.loss, feed_dict=feed_dict)
         return {'loss': loss, 'learning_rate': feed_dict[self.learning_rate_ph]}
 
+    def train_on_batch(self, features: List[InputFeatures], y: Union[List[int], List[List[int]]] = None) -> Dict:
```
Comment (dilyararimovna) [inline deeppavlov/models/bert/bert_classifier.py:236]: ты добавляешь новую фичу для тренировки, при это я не вижу оставленной возможности тренировтаься как раньше. Должна быть проверка "что если использовать gradient accumulation, то одним образом тренировка, если не использовать, то другим способом.

File: deeppavlov/models/bert/bert_ranker.py [hunk 4]
```diff
@@ -86,18 +79,9 @@ def __call__(self, features_li: List[List[InputFeatures]]) -> Union[List[int], L
...
-            input_masks = [f.input_mask for f in features]
-            input_type_ids = [f.input_type_ids for f in features]
-
-            feed_dict = self._build_feed_dict(input_ids, input_masks, input_type_ids)
-            if not self.return_probas:
-                pred = self.sess.run(self.y_predictions, feed_dict=feed_dict)
-            else:
-                pred = self.sess.run(self.y_probas, feed_dict=feed_dict)
+            pred = super()(features)
             predictions.append(pred[:, 1])
```
Comment (dilyararimovna) [inline deeppavlov/models/bert/bert_ranker.py:85]: тут раньше была возможно вероятности или лейблы выдавать,сейчас только один вариант остался

File: deeppavlov/core/trainers/nn_trainer.py [hunk 5]
```diff
@@ -90,6 +93,7 @@ class NNTrainer(FitTrainer):
 
     def __init__(self, chainer_config: dict, *, 
                  batch_size: int = 1,
+                 gradient_accumulation_steps: int = 1,
```
Comment (dilyararimovna) [inline deeppavlov/core/trainers/nn_trainer.py:96]: зачем? нигде не используется дальше в тренере

**Review by dilyararimovna — CHANGES_REQUESTED** [review]

File: deeppavlov/models/bert/bert_ranker.py (hunk 4 above)
Comment (dimakarp1996) [reply deeppavlov/models/bert/bert_ranker.py:85]: Вызывается функция суперкласса, у которого остаётся аргумент self.return_probas и который может сам выдавать вероятности или лейблы. Т.е ничего не пропало, а только перекочевало туда.

File: deeppavlov/models/bert/bert_classifier.py (hunk 3 above)
Comment (dimakarp1996) [reply deeppavlov/models/bert/bert_classifier.py:236]: Fixed

File: deeppavlov/models/bert/bert_classifier.py (hunk 1 above)
Comment (dimakarp1996) [reply deeppavlov/models/bert/bert_classifier.py:221]: Вернул

File: deeppavlov/core/trainers/utils.py (hunk 2 above)
Comment (dimakarp1996) [reply deeppavlov/core/trainers/utils.py:20]: здесь == где? В импорты пустые строки вернул

File: deeppavlov/core/trainers/nn_trainer.py (hunk 5 above)
Comment (dimakarp1996) [reply deeppavlov/core/trainers/nn_trainer.py:96]: ок, убрал

File: deeppavlov/models/bert/bert_ranker.py (hunk 4 above)
Comment (dilyararimovna) [reply deeppavlov/models/bert/bert_ranker.py:85]: ок

File: deeppavlov/models/bert/bert_classifier.py (hunk 3 above)
Comment (dilyararimovna) [reply deeppavlov/models/bert/bert_classifier.py:245]: зачем делать две функции? мне кажется, проще код из train_on_batch_old перенести сюда

**Review by dilyararimovna — CHANGES_REQUESTED** [review]
поправь код стайл - пробелов местами не хватает

File: deeppavlov/core/trainers/utils.py (hunk 2 above)
Comment (dimakarp1996) [reply deeppavlov/core/trainers/utils.py]: Ok

File: deeppavlov/models/bert/bert_classifier.py [hunk 6]
```diff
@@ -13,8 +13,9 @@
 # limitations under the License.
 
 from logging import getLogger
-from typing import List, Dict, Union
+from typing import List, Dict, Union, Optional
 
+import math
```
[the remark this answers is not in the document]
Comment (dimakarp1996) [reply deeppavlov/models/bert/bert_classifier.py:17]: Ok

File: deeppavlov/models/bert/bert_classifier.py [hunk 7]
```diff
@@ -168,21 +176,14 @@ def _init_optimizer(self):
                                                initializer=tf.constant_initializer(0), trainable=False)
             # default optimizer for Bert is Adam with fixed L2 regularization
             if self.optimizer is None:
```
Comment (yurakuratov) [inline deeppavlov/models/bert/bert_classifier.py:178]: What will happen if optimizer is specified?

File: deeppavlov/models/bert/bert_classifier.py [hunk 8]
```diff
@@ -168,21 +176,14 @@ def _init_optimizer(self):
...
-                                                  beta_2=0.999,
-                                                  epsilon=1e-6,
-                                                  exclude_from_weight_decay=["LayerNorm", "layer_norm", "bias"]
-                                                  )
-            else:
-                self.train_op = self.get_train_op(self.loss, learning_rate=self.learning_rate_ph)
-
-            if self.optimizer is None:
-                new_global_step = self.global_step + 1
-                self.train_op = tf.group(self.train_op, [self.global_step.assign(new_global_step)])
```
Comment (yurakuratov) [inline deeppavlov/models/bert/bert_classifier.py:185]: These lines should be used when `AdamWeightDecayOptimizer` is used.

File: deeppavlov/models/bert/bert_classifier.py [hunk 9]
```diff
@@ -168,21 +176,14 @@ def _init_optimizer(self):
                                                initializer=tf.constant_initializer(0), trainable=False)
             # default optimizer for Bert is Adam with fixed L2 regularization
             if self.optimizer is None:
-
-                self.train_op = self.get_train_op(self.loss, learning_rate=self.learning_rate_ph,
```
Comment (yurakuratov) [inline deeppavlov/models/bert/bert_classifier.py:172]: check get_train_op function implementation, you lose all features that are implemented in it by using raw optimizer

File: deeppavlov/models/bert/bert_classifier.py [hunk 10]
```diff
@@ -200,7 +201,27 @@ def _build_feed_dict(self, input_ids, input_masks, token_types, y=None):
...
+        Args:
+            features: batch of InputFeatures
+
+        Returns:
+            list of input feature batches, with the size equal to the gradient_accumulation_steps
+
+        """
+
+        num_parts = self.gradient_accumulation_steps
+        assert num_parts > 0
```
Comment (yurakuratov) [inline deeppavlov/models/bert/bert_classifier.py:218]: This could be checked in `__init__` method and raise exception in case of illegal values

**Review by yurakuratov — CHANGES_REQUESTED** [review]
in addition to comments above:
- report results of testing on SST-2
- gradient accumulation is a very useful feature. It should be added not only to BERT classification models (check if it possible to move impl to TFModel)
- can you report checks that loss and gradients are equal in cases when gradient_accumulations_steps = 1 and gradient_accumulations_steps > 1 (they should be almost equal for large enough batches)
- add at least one test to check gradient_accumulation (test configuration file with enabled gradient accumulation)

File: deeppavlov/models/bert/bert_classifier.py (hunk 8 above)
Comment (dimakarp1996) [reply deeppavlov/models/bert/bert_classifier.py:185]: Ок

File: deeppavlov/models/bert/bert_classifier.py (hunk 7 above)
Comment (dimakarp1996) [reply deeppavlov/models/bert/bert_classifier.py:181]: Ок

File: deeppavlov/models/bert/bert_classifier.py (hunk 10 above)
Comment (dimakarp1996) [reply deeppavlov/models/bert/bert_classifier.py]: Ок

File: deeppavlov/models/bert/bert_classifier.py [hunk 11]
```diff
@@ -200,7 +201,27 @@ def _build_feed_dict(self, input_ids, input_masks, token_types, y=None):
 
         return feed_dict
 
-    def train_on_batch(self, features: List[InputFeatures], y: Union[List[int], List[List[int]]]) -> Dict:
+    def split(self, features: List):
```
[the remark this answers is not in the document]
Comment (dimakarp1996) [reply deeppavlov/models/bert/bert_classifier.py]: Ок

**Comment by dimakarp1996** [comment]
I am currently running testing on SST-2. Test config will be added. Check will be done manually. I doubt if it is possible to move implementation, as train_on_batch method is not defined in the base class and therefore has to be redefined in all subclasses.

**Comment by dimakarp1996** [comment]
1)Added two test configs
2)Metrics in a config without gradient accumulation:
{"train": {"eval_examples_count": 60614, "metrics": {"accuracy": 0.9889, "f1_macro": 0.9887}, "time_spent": "0:05:15"}}
{"valid": {"eval_examples_count": 6735, "metrics": {"accuracy": 0.9535, "f1_macro": 0.9528}, "time_spent": "0:00:35"}}
{"test": {"eval_examples_count": 872, "metrics": {"accuracy": 0.9278, "f1_macro": 0.9276}, "time_spent": "0:00:05"}}
Metrics in a config with gradient accumulation:
{"train": {"eval_examples_count": 60614, "metrics": {"accuracy": 0.9898, "f1_macro": 0.9897}, "time_spent": "0:05:06
{"valid": {"eval_examples_count": 6735, "metrics": {"accuracy": 0.9523, "f1_macro": 0.9516}, "time_spent": "0:00:34"
{"test": {"eval_examples_count": 872, "metrics": {"accuracy": 0.9243, "f1_macro": 0.9242}, "time_spent": "0:00:05"}}

**Comment by dimakarp1996** [comment]
No gradient accumulation(batch size 32):
{"train": {"eval_examples_count": 60614, "metrics": {"accuracy": 0.9889, "f1_macro": 0.9887}, "time_spent": "0:05:15"}}
{"valid": {"eval_examples_count": 6735, "metrics": {"accuracy": 0.9535, "f1_macro": 0.9528}, "time_spent": "0:00:35"}}
{"test": {"eval_examples_count": 872, "metrics": {"accuracy": 0.9278, "f1_macro": 0.9276}, "time_spent": "0:00:05"}}
